pkg/machines/components/vcpuModal.jsx

```diff
@@ -0,0 +1,192 @@
+        if (num % i === 0) {
+            divs.push(i);
+        }
+    }
+
+    if (num > 1) {
+        divs.push(num);
+    }
+
+    return divs;
+}
+
+const Select = function ({ id, items, onChange, value }) {
+    return ( <SelectComponent.Select id={id} initial={value} onChange={onChange}>
+        {items.map((t) => (
+            <SelectComponent.SelectEntry data={t}>{t}</SelectComponent.SelectEntry>
+        ))}
+    </SelectComponent.Select>);
+}
+
+class VCPUModalBody extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            max: props.vcpus.max || 1,
+            count: parseInt(props.vcpus.count) || 1,
+            sockets: props.cpu.topology.sockets || 1,
+            threads: props.cpu.topology.threads || 1,
+            cores: props.cpu.topology.cores || 1
+        };
+        this.onMaxChange = this.onMaxChange.bind(this);
+        this.onCountSelect = this.onCountSelect.bind(this);
+        this.onSocketChange = this.onSocketChange.bind(this);
+        this.onThreadsChange = this.onThreadsChange.bind(this);
+        this.onCoresChange = this.onCoresChange.bind(this);
+        props.onChange(this.state);
+    }
+
+    componentWillUpdate (nextProps, nextState) {
+        this.props.onChange(nextState);
+    }
+
+    onMaxChange (e) {
+        const maxHypervisor = parseInt(this.props.hypervisorMax);
+        let maxValue = parseInt(e.target.value);
+        if (isNaN(maxValue) || maxValue <= 0) {
+            maxValue = 1;
+        }
+        if (maxValue > maxHypervisor) {
+            maxValue = maxHypervisor;
+        }
+        if (maxValue < this.state.count) {
+            this.setState({ count: maxValue });
+        }
+        const state = { max: maxValue, sockets: this.state.sockets, cores: this.state.cores };
+        let divs = dividers(state.max);
+        if (divs.indexOf(this.state.sockets) === -1 || (this.props.cpu.topology.sockets || 1) === this.state.sockets) {
+            state.sockets = divs[divs.length - 1];
+        }
+        divs = dividers(state.max / state.sockets);
+        if (divs.indexOf(this.state.cores) === -1) {
+            state.cores = divs[divs.length - 1];
+        }
+        state.threads = state.max / ( state.cores * state.sockets );
+        this.setState(state);
+    }
+
+    onCountSelect (e) {
+        let value = parseInt(e.target.value);
+        if (isNaN(value) || value <= 0) {
+            value = 1;
+        } else {
+            if (value > this.state.max) {
+                value = this.state.max;
+            }
+        }
+        this.setState({ count: parseInt(value) });
+    }
+
+    onSocketChange (value) {
+        const state = { sockets: this.state.sockets, cores: this.state.cores };
+        state.sockets = parseInt(value);
+        let divs = dividers(this.state.max / parseInt(value));
+        if (divs.indexOf(this.state.cores) === -1) {
+            state.cores = divs[divs.length - 1];
+        }
```

Comment:
Likewise, this could use a comment about the assumptions and logic. Same thing for threads and cores below.